File: lib/day1/day1.ex
defmodule Advent.Day1 do
  @input File.read!("#{__DIR__}/input.txt") |> String.split("\n")

  def problem1 do
    find_frequency(0, @input)
  end

  def problem2 do
    find_repeated(0, @input)
  end

  def find_frequency(start, inputs) do
    Enum.reduce(inputs, start, fn i, acc ->
      next(acc, i)
    end)
  end

  def find_repeated(start, input, results \\ %{0 => true}) do
    Enum.reduce_while(input, {start, results}, fn i, {acc, res} ->
      acc = next(acc, i)

      if res[acc] do
        {:halt, {:found, acc}}
      else
        {:cont, {acc, Map.put(res, acc, true)}}
      end
    end)
    |> case do
      {:found, val} -> val
      {acc, r} -> find_repeated(acc, input, r)
    end
  end

  defp next(acc, num) do
    acc + String.to_integer(num)
  end
end

File: golang/day1/day1.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func FindFrequency(inputs []string) (acc int) {
	for _, val := range inputs {
		ival, _ := strconv.Atoi(val)
		acc += ival
	}
	return
}

func FindRepeating(lines []string, acc int, foundList map[int]bool) int {
	foundList[acc] = true
	for _, s := range lines {
		val, _ := strconv.Atoi(s)
		acc += val
		if foundList[acc] {
			return acc
		}
		foundList[acc] = true
	}
	return FindRepeating(lines, acc, foundList)
}

func valInSlice(a int, list []int) bool {
	for _, b := range list {
		if b == a {
			return true
		}
	}
	return false
}

func main() {
	content, _ := ioutil.ReadFile("../../inputs/day1.txt")
	lines := strings.Split(string(content), "\n")

	fmt.Println("problem1:", FindFrequency(lines))
	fmt.Println("problem2:", FindRepeating(lines, 0, map[int]bool{}))
}
